port-chooser/src/main/java/org/terracotta/utilities/test/io/WindowsSpecialFolder.java

```diff
@@ -0,0 +1,161 @@
+
+  WindowsSpecialFolder(String identifier, boolean create) {
+    requireNonNull(identifier, "identifier");
+    this.accessor = LazyProperty.lazily(() -> getSpecialFolder(identifier, create));
+  }
+
+  /**
+   * Gets the {@code Path} assigned to the identified special folder.
+   * @return the special folder {@code Path}
+   * @throws IOException if an error is raised while attempting to determine the
+   *      {@code Path} for the special folder
+   */
+  public Path get() throws IOException {
+    return accessor.get();
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WindowsSpecialFolder.class);
+
+  /**
+   * Gets the {@link Path} to the identified Windows Special Folder.
+   *
+   * @param specialFolderId the special folder identifier
+   * @param create          if {@code true}, attempts to create the folder if it does not exist
+   * @return the path to the special folder; an empty path is returned if the folder does not exist
+   * @throws IOException if the special folder value cannot be determined
+   */
+  private static Path getSpecialFolder(String specialFolderId, boolean create) throws IOException {
+    String[] command = new String[] {
+        "powershell.exe",
+        "-NoLogo",
+        "-NoProfile",
+        "-NonInteractive",
+        "-Command",
+        "&{$ErrorActionPreference = 'Stop'; " +
+            "[environment]::getfolderpath('" + specialFolderId + "'" + (create ? ", 'create'" : "") + ")}" };
+
+    String specialFolder;
+    Shell.Result result;
+    try {
+      result = Shell.execute(Shell.Encoding.CHARSET, command);
+    } catch (IOException e) {
+      LOGGER.error("Unable to determine special folder for {}; {} failed",
+          specialFolderId, Arrays.toString(command), e);
+      throw e;
+    }
+    if (result.exitCode() == 0) {
+      specialFolder = result.lines().get(0);
+    } else {
+      SpecialFolderException exception =
+          new SpecialFolderException(specialFolderId, result.lines(), result.exitCode());
+      LOGGER.error("Unable to determine special folder for {}", specialFolderId, exception);
+      throw exception;
+    }
+
+    return Paths.get(specialFolder);
+  }
+
+  /**
+   * A <i>derived on first reference</i> value holder.
+   * @param <T> the value type
+   */
+  private static final class LazyProperty<T> {
+    public static <T> LazyProperty<T> lazily(ThrowingSupplier<T> from) {
+      return new LazyProperty<>(from);
+    }
+
+    private final AtomicReference<T> reference = new AtomicReference<>();
+    private final ThrowingSupplier<T> supplier;
+
+    private LazyProperty(ThrowingSupplier<T> supplier) {
+      this.supplier = supplier;
+    }
+
+    public T get() throws IOException {
+      T t;
+      do {
+        t = reference.get();
+      } while (t == null && !reference.compareAndSet(null, t = supplier.get()));
+      return t;
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T> {
```

Comment:
Could we not just use `Callable` instead?

Reply:
I could but I really wanted just an `IOException` out of `get`.